pkg/middleware/subpath_redirect.go

```diff
@@ -0,0 +1,20 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/web"
+)
+
+// Redirects URLs that are missing the configured subpath to an URL that contains the subpath.
+func SubPathRedirect(cfg *setting.Cfg) web.Handler {
+	return func(res http.ResponseWriter, req *http.Request, c *web.Context) {
+		if !strings.HasPrefix(req.RequestURI, cfg.AppSubURL) {
```

Comment:
Maybe we should avoid redirecting requests hitting the `/api/*` to avoid breaking any setup that uses the API for automation. 

Most http libraries does follow redirect but we ran into this problem when we built the hg-gateway where a customers used some python library that doesn't follow redirects.

Reply:
@bergquist yes, agree we can ignore urls starting with /api